pkg/tso/keyspace_group_manager.go

```diff
@@ -475,59 +484,169 @@ func (kgm *KeyspaceGroupManager) updateKeyspaceGroup(group *endpoint.KeyspaceGro
+		kgm.kgs[group.ID].Store(group)
 		kgm.ams[group.ID].Store(NewAllocatorManager(kgm.ctx, group.ID, participant, tsRootPath, storage, kgm.cfg, true))
-		kgm.ksgs[group.ID].Store(group)
 	} else {
 		// Not assigned to me. If this host/pod owns this keyspace group, it should resign.
 		kgm.deleteKeyspaceGroup(group.ID)
 	}
 }
 
+// updateKeyspaceGroupMembership updates the keyspace lookup table for the given keyspace group.
+func (kgm *KeyspaceGroupManager) updateKeyspaceGroupMembership(
+	groupID uint32, oldKeyspaces, newKeyspaces []uint32,
+	oldKeyspaceLookupTable map[uint32]struct{},
+) map[uint32]struct{} {
+	oldLen := len(oldKeyspaces)
+	newLen := len(newKeyspaces)
+
+	// Sort the keyspaces in ascending order
+	sort.Slice(newKeyspaces, func(i, j int) bool {
+		return newKeyspaces[i] < newKeyspaces[j]
+	})
+
+	// Mostly, the membership has no change, so we optimize for this case.
+	sameMembership := true
+	if oldLen != newLen {
+		sameMembership = false
+	} else {
+		for i := 0; i < oldLen; i++ {
+			if oldKeyspaces[i] != newKeyspaces[i] {
+				sameMembership = false
+				break
+			}
+		}
+	}
+
+	var newKeyspaceLookupTable map[uint32]struct{}
+
+	if sameMembership {
+		// The keyspace group membership is not changed, so we reuse the old one.
+		newKeyspaceLookupTable = oldKeyspaceLookupTable
+	} else {
+		// The keyspace group membership is changed, so we update the keyspace lookup table.
+		newKeyspaceLookupTable = make(map[uint32]struct{})
+		for i, j := 0, 0; i < oldLen || j < newLen; {
+			if i < oldLen && j < newLen && oldKeyspaces[i] == newKeyspaces[j] {
+				newKeyspaceLookupTable[newKeyspaces[j]] = struct{}{}
+				i++
+				j++
+			} else if i < oldLen && j < newLen && oldKeyspaces[i] < newKeyspaces[j] || j == newLen {
+				kgm.keyspaceLookupTable.Delete(oldKeyspaces[i])
+				i++
+			} else {
+				newKeyspaceLookupTable[newKeyspaces[j]] = struct{}{}
+				kgm.keyspaceLookupTable.Store(newKeyspaces[j], groupID)
+				j++
+			}
+		}
+	}
+
+	return newKeyspaceLookupTable
+}
+
+func (kgm *KeyspaceGroupManager) buildKeyspaceLookupTable(groupID uint32, keyspaces []uint32) map[uint32]struct{} {
+	keyspaceLookupTable := make(map[uint32]struct{})
+	for _, kid := range keyspaces {
+		keyspaceLookupTable[kid] = struct{}{}
+		kgm.keyspaceLookupTable.Store(kid, groupID)
+	}
+	return keyspaceLookupTable
+}
+
 // deleteKeyspaceGroup deletes the given keyspace group.
-func (kgm *KeyspaceGroupManager) deleteKeyspaceGroup(id uint32) {
-	kgm.ksgs[id].Store(nil)
-	am := kgm.ams[id].Swap(nil)
-	if am == nil {
-		return
+func (kgm *KeyspaceGroupManager) deleteKeyspaceGroup(groupID uint32) {
+	kg := kgm.kgs[groupID].Swap(nil)
+	if kg != nil {
+		for _, kid := range kg.Keyspaces {
+			kgm.keyspaceLookupTable.CompareAndDelete(kid, kg.ID)
+		}
 	}
-	am.close()
-	log.Info("deleted keyspace group", zap.Uint32("keyspace-group-id", id))
+	am := kgm.ams[groupID].Swap(nil)
+	if am != nil {
+		am.close()
+	}
```

Comment:
Should we make it atomic?

Reply:
discussed offline. leave what it is for now, and we may improve it in the future.

Reply:
And maybe we can improve it later.

Reply:
Very tiny unreliable/inconsistent window and we can just let client retry to get over it. After all, there are other unreliable/inconsistency windows during kg-split/keyspace-movement/kg-movement which can't be avoided. It won't cause incorrect result for tso requests, because the timeline split problem will be specifically handled by kg-split/keyspace-movement and there is no time split problem because of primary election among replicas within a kg.